stacks-signer/src/runloop.rs

```diff
@@ -937,21 +1069,98 @@ impl<C: Coordinator> SignerRunLoop<Vec<OperationResult>, RunLoopCommand> for Run
     }
 }
 
+/// When this threshold is reached, the signer will use the NACKed CoordinatorMetadata if that data
+/// is more updated than its own chain view
+fn get_nack_threshold(total_signers: u32) -> u32 {
+    (total_signers + 1) / 2
+}
+
+/// Clean up outdated coordinator_metadata NACKs that have a stale or equivalent chain view compared to the current signer
+fn cleanup_outdated_coordinator_nacks(
+    nack_messages_received: &mut HashMap<CoordinatorMetadata, HashSet<u32>>,
+    current_observed_height: u64,
+) {
+    let outdated_metadata_to_remove: Vec<CoordinatorMetadata> = nack_messages_received
+        .iter()
+        .filter(|(coordinator_metadata, _)| {
+            coordinator_metadata.stacks_block_height <= current_observed_height
+        })
+        .map(|(coordinator_metadata, _)| *coordinator_metadata)
+        .collect();
+
+    for coordinator_metadata in outdated_metadata_to_remove {
+        nack_messages_received.remove(&coordinator_metadata);
+    }
+}
+
+/// Returns the coordinatorMetadata broadcasted by most signers if it meets the NACK threshold
+fn get_lead_coordinator_by_nack(
+    nack_messages_received: &HashMap<CoordinatorMetadata, HashSet<u32>>,
+    nack_threshold: &u32,
+) -> Option<CoordinatorMetadata> {
+    nack_messages_received
+        .iter()
+        .filter(|(_, signers)| signers.len() >= *nack_threshold as usize)
+        .max_by_key(|(_, signers)| signers.len())
+        .map(|(coordinator_metadata, _)| *coordinator_metadata)
+}
+
+/// Create coordinator Info with fake consensus hash and block height to not block signers if they can't calculate
+/// coordinator properly even after retries. This will result to NACKs and will eventually get ignored/overthrown
+fn get_default_coordinator_info(signer_id: u32, public_key: PublicKey) -> CoordinatorInfo {
+    let default_coordinator_metadata = get_default_coordinator_metadata();
+    CoordinatorInfo::new(signer_id, public_key, default_coordinator_metadata)
+}
+
+fn get_default_coordinator_metadata() -> CoordinatorMetadata {
```

Comment:
Nit: instead of having a function for this, can you just implement `Default` for `CoordinatorMetadata`?